### PR TITLE
BUG: GroupBy.ffill()/bfill() do not return NaN values for NaN groups

### LINKED ISSUE #34725: BUG: fillna(method="ffill") and ffill() on DataFrameGroupBy gives different results
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here

import numpy as np
import pandas as pd
df = pd.DataFrame({'a':[1, 1, np.NaN, np.NaN], 'b':[10, np.NaN, 40, np.NaN]})
```
```
df
     a     b
0  1.0  10.0
1  1.0   NaN
2  NaN  40.0
3  NaN   NaN
```

```
# Case 1 - Forward fill using fillna
df.groupby(["a"]).fillna(method="ffill")

df
      b
0  10.0
1  10.0
2   NaN
3   NaN
```

```
# Case 2 - Forward fill using ffill
df.groupby(["a"]).ffill()

df
      b
0  10.0
1  10.0
2  40.0
3  40.0
```

#### Problem description
`df.groupby(["a"]).ffill()` is not ignoring the NaN group while performing forward fill. 

#### Expected Output
`df.groupby(["a"]).ffill()` should give the same output as `df.groupby(["a"]).fillna(method="ffill")` since NaN groups are ignored as per https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html#na-values-in-groupby.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 16.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : en_US.UTF-8
pandas           : 1.0.4
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 44.0.0.post20200106
Cython           : None
pytest           : 5.4.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : 0.17.0
pytables         : None
pytest           : 5.4.3
pyxlsb           : None
s3fs             : 0.4.2
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : 0.8.6
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by smithto1**
- [x] closes #34725
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

See #34725 for a good description of the problem. 

This fixes the issue in the tidiest way I could find. 

One test is added that covers all relevant cases I could think of. The comments in the test can be removed, they are just there to make it easier for the reviewer to follow as the test is a bit complicated. 

This also fixes the copy-pastable example from the #34725.

### CODE REVIEW

**Comment by rhshadrach** [comment]
Is there a way to fix `group_fillna_indexer` rather than correct the results with post-processing? Or is there a reason _not_ to do this?

File: pandas/tests/groupby/transform/test_transform.py [hunk 1]
```diff
@@ -977,6 +977,76 @@ def test_ffill_bfill_non_unique_multilevel(func, expected_status):
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.parametrize("dropna", [True, False])
+@pytest.mark.parametrize("limit", [None, 1])
+@pytest.mark.parametrize("method", ["ffill", "bfill", "pad", "backfill"])
+@pytest.mark.parametrize("by", ["grp1", ["grp1"], ["grp1", "grp2"]])
+@pytest.mark.parametrize("has_nan", [[], ["grp1"], ["grp1", "grp2"]])
+def test_pad_handles_nan_groups(dropna, limit, method, by, has_nan):
```
Comment (WillAyd) [inline pandas/tests/groupby/transform/test_transform.py:985]: Can you simplify this test? Things should be modular and there shouldn't really be a need to add long comments to the tests, so I think this is a good sign that this should be simplified

**Review by WillAyd — CHANGES_REQUESTED** [review]
Agreed with @rhshadrach comment here - this should be fixed directly in the underlying function

File: doc/source/whatsnew/v1.1.3.rst [hunk 2]
```diff
@@ -58,6 +58,7 @@ Bug fixes
 - Bug in :meth:`Series.astype` showing too much precision when casting from ``np.float32`` to string dtype (:issue:`36451`)
 - Bug in :meth:`Series.isin` and :meth:`DataFrame.isin` when using ``NaN`` and a row length above 1,000,000 (:issue:`22205`)
 - Bug in :func:`cut` raising a ``ValueError`` when passed a :class:`Series` of labels with ``ordered=False`` (:issue:`36603`)
+- Bug in :meth:`DataFrameGroupBy.ffill` where a ``NaN`` group would return foward-filled values instead of ``NaN`` when ``dropna=True`` (:issue:`34725`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.1.3.rst:61]: Can you also move this to 1.2?  I don't believe this is a regression in the 1.1 series

**Comment by smithto1** [comment]
> Is there a way to fix `group_fillna_indexer` rather than correct the results with post-processing? Or is there a reason _not_ to do this?

@rhshadrach This is fixed in `group_fillna_indexer` now. It is much cleaner and faster.

File: pandas/tests/groupby/transform/test_transform.py (hunk 1 above)
Comment (smithto1) [reply pandas/tests/groupby/transform/test_transform.py]: @WillAyd The test has been simplified. Implementing the change in the underlying cython function touches fewer things and can get by with a simpler test.

File: doc/source/whatsnew/v1.1.3.rst (hunk 2 above)
Comment (smithto1) [reply doc/source/whatsnew/v1.1.3.rst]: Done

File: pandas/_libs/groupby.pyx [hunk 3]
```diff
@@ -389,6 +390,8 @@ def group_fillna_indexer(ndarray[int64_t] out, ndarray[int64_t] labels,
             else:  # reset items when not missing
                 filled_vals = 0
                 curr_fill_idx = idx
+            if dropna and labels[idx] == -1:
+                curr_fill_idx = -1
```
Comment (rhshadrach) [inline pandas/_libs/groupby.pyx:394]: I believe you can move these two lines to the top of the if-else, and change the `if mask[idx] == 1:` to `elif mask[idx] == 1`. This way we're not executing code needlessly.

File: doc/source/whatsnew/v1.2.0.rst [hunk 4]
```diff
@@ -413,6 +413,7 @@ Groupby/resample/rolling
 - Bug in :meth:`Rolling.sum()` returned wrong values when dtypes where mixed between float and integer and axis was equal to one (:issue:`20649`, :issue:`35596`)
 - Bug in :meth:`Rolling.count` returned ``np.nan`` with :class:`pandas.api.indexers.FixedForwardWindowIndexer` as window, ``min_periods=0`` and only missing values in window (:issue:`35579`)
 - Bug where :class:`pandas.core.window.Rolling` produces incorrect window sizes when using a ``PeriodIndex`` (:issue:`34225`)
+- Bug in :meth:`DataFrameGroupBy.ffill` where a ``NaN`` group would return foward-filled values instead of ``NaN`` when ``dropna=True`` (:issue:`34725`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.2.0.rst:416]: bfill was also impacted by this fix, right? It should be mentioned here as well.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looking really good, two minor requests.

File: pandas/_libs/groupby.pyx (hunk 3 above)
Comment (smithto1) [reply pandas/_libs/groupby.pyx]: Done

File: doc/source/whatsnew/v1.2.0.rst (hunk 4 above)
Comment (smithto1) [reply doc/source/whatsnew/v1.2.0.rst]: Done.

**Comment by smithto1** [comment]
@rhshadrach @WillAyd  Comments have been addressed. Can you please have another look?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looks good. Can you move tests to tests.groupby.test_missing - ref #36326

**Comment by smithto1** [comment]
> Looks good. Can you move tests to tests.groupby.test_missing - ref #36326

@rhshadrach Done.

**Review by rhshadrach — APPROVED** [review]

**Comment by rhshadrach** [comment]
@smithto1 CI failure due to import order in unrelated files; I think merging master will fix.

**Comment by smithto1** [comment]
> @smithto1 CI failure due to import order in unrelated files; I think merging master will fix.

@rhshadrach I think that is all fixed now.

**Comment by rhshadrach** [comment]
@smithto1 Got a new conflict that needs resolved.

**Review by jreback — APPROVED** [review]
lgtm can merge on green.

**Comment by jreback** [comment]
thanks @smithto1

**Pull request merged**